dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -557,35 +559,130 @@ private ISKFunction GetFunction(FunctionView targetFunction)
+        }
+        else
+        {
+            result = new List<FunctionView>();
+
+            // Remember functions in memory so that they can be searched.
+            await RememberFunctionsAsync(this.Config.Memory, availableFunctions, cancellationToken).ConfigureAwait(false);
+
+            // Search for functions that match the semantic query.
+            var memories = this.Config.Memory.SearchAsync(
+                PlannerMemoryCollectionName,
+                semanticQuery!,
+                this.Config.MaxRelevantFunctions,
+                this.Config.RelevancyThreshold.Value,
+                cancellationToken: cancellationToken);
 
-            var excludedSkills = this.Config.ExcludedSkills ?? new();
-            var excludedFunctions = this.Config.ExcludedFunctions ?? new();
+            // Add functions that were found in the search results.
+            result.AddRange(await this.GetRelevantFunctionsAsync(availableFunctions, memories, cancellationToken).ConfigureAwait(false));
 
-            var availableFunctions =
-                functionsView
-                    .Where(s => !excludedSkills.Contains(s.SkillName) && !excludedFunctions.Contains(s.Name))
-                    .OrderBy(x => x.SkillName)
-                    .ThenBy(x => x.Name);
+            // Add any missing functions that were included but not found in the search results.
+            var missingFunctions = this.Config.IncludedFunctions
+                .Except(result.Select(x => x.Name))
+                .Join(availableFunctions, f => f, af => af.Name, (_, af) => af);
+
+            result.AddRange(missingFunctions);
+        }
+
+        return result
+            .OrderBy(x => x.SkillName)
+            .ThenBy(x => x.Name);
+    }
+
+    private async Task<IEnumerable<FunctionView>> GetRelevantFunctionsAsync(
+        IEnumerable<FunctionView> availableFunctions,
+        IAsyncEnumerable<MemoryQueryResult> memories,
+        CancellationToken cancellationToken = default)
+    {
+        var relevantFunctions = new ConcurrentBag<FunctionView>();
+        await foreach (var memoryEntry in memories.WithCancellation(cancellationToken))
+        {
+            var function = availableFunctions.FirstOrDefault(x => ToFullyQualifiedName(x) == memoryEntry.Metadata.Id);
+            if (function != null)
+            {
+                if (this._logger?.IsEnabled(LogLevel.Debug) == true)
+                {
+                    this._logger?.LogDebug("Found relevant function. Relevance Score: {0}, Function: {1}", memoryEntry.Relevance, ToFullyQualifiedName(function));
+                }
 
-            return Task.FromResult(availableFunctions);
+                relevantFunctions.Add(function);
+            }
         }
 
-        return this.Config.GetAvailableFunctionsAsync(this.Config, null, cancellationToken);
+        return relevantFunctions;
+    }
+
+    /// <summary>
+    /// Saves all available functions to memory.
+    /// </summary>
```

Comment:
After the planners are merged, it'll be easier to reduce the code duplication between this and sequential planner extensions.